src/Authentication.ts

```diff
@@ -1,231 +1,365 @@
+    validationAdapter.initialize(this)
+    this.validationAdapter = validationAdapter
+  }
+
+  /**
+   * Getter for the storage adapter.
+   */
+  public get StorageAdapter (): StorageAdapterInterface {
+    return this.storageAdapter
+  }
+
+  /**
+   * Setter for the storage adapter.
+   */
+  public set StorageAdapter (storageAdapter: StorageAdapterInterface) {
+    storageAdapter.initialize(this)
+    this.storageAdapter = storageAdapter
+  }
+
+  /**
+   * Getter for the cache adapter.
+   */
+  public get CacheAdapter (): CacheAdapterInterface {
+    return this.cacheAdapter
+  }
+
+  /**
+   * Setter for the cache adapter.
+   */
+  public set CacheAdapter (cacheAdapter: CacheAdapterInterface) {
+    cacheAdapter.initialize(this)
+    this.cacheAdapter = cacheAdapter
+  }
+
+  /**
+   * Getter for the global auth config object.
+   */
+  public get GlobalAuthConfig (): object {
+    return this.globalAuthConfig
+  }
+
+  /**
+   * Getter for the two-factor providers.
+   */
+  public get TwoFactorProviders (): Map<string, TwoFactorProviderInterface> {
+    return this.twoFactorProviders
+  }
+
+  /**
+   * Setter for the two-factor providers.
+   */
+  public set TwoFactorProviders (twoFactorProviders: Map<string, TwoFactorProviderInterface>) {
+    this.twoFactorProviders = twoFactorProviders
+  }
+
+  /**
+   * Setter method used to add a new two-factor provider.
+   *
+   * @param {TwoFactorProviderInterface} payload
+   */
+  addTwoFactorProvider (payload: TwoFactorProviderInterface): void { // PITANJE da ostavim da prepise vrijednost ako vec postoji ili da radim provjeru
+    this.twoFactorProviders.set(payload.provider, payload)
+  }
+
+  /**
+   * Getter method used to fetch a certain two-factor provider.
+   *
+   * @param {string} provider
+   * @return {TwoFactorProviderInterface}
+   * @throws
+   */
+  getTwoFactorProvider (provider: string): TwoFactorProviderInterface {
+    const result = this.twoFactorProviders.get(provider)
+
+    if (result === undefined) {
+      throw new AuthenticationError('provider not found', { name: 'AuthenticationError', statusCode: 500 })
+      // PITANJE da napravim try catch i throwam kako bi handleError() metoda rijesila error
+    }
+
+    return result
+  }
+
+  /**
+   * Register user.
+   *
+   * @param {RegistrationData} payload
+   * @return {Promise<{ user: AuthenticableUser, twoFactorUser?: AuthenticableTwoFactorUser, qrCode?: string }>}
+   * @throws
+   */
+  async register (payload: RegistrationData): Promise<{ user: AuthenticableUser, twoFactorUser?: AuthenticableTwoFactorUser, qrCode?: string }> {
+    try {
+      this.validationAdapter.registration(payload)
+
+      const user = await this.storageAdapter.register(payload)
+      const result: { user: AuthenticableUser, twoFactorUser?: AuthenticableTwoFactorUser, qrCode?: string } = { user }
+
+      if (payload.twoFactorAuthenticationProvider !== undefined) {
```

Comment:
what if `payload.twoFactorAuthenticationProvider` is `null`?

Reply:
Added the null check